graphql_compiler/schema/schema_info.py

```diff
@@ -1,13 +1,166 @@
 # Copyright 2019-present Kensho Technologies, LLC.
 from collections import namedtuple
+from dataclasses import dataclass
+from functools import partial
+from typing import Dict, Optional
 
+from graphql.type import GraphQLSchema
 from graphql.type.definition import GraphQLInterfaceType, GraphQLObjectType
 import six
 import sqlalchemy
+from sqlalchemy.dialects.mssql import dialect as mssql_dialect
+from sqlalchemy.dialects.mysql import dialect as mysql_dialect
+from sqlalchemy.dialects.postgresql import dialect as postgresql_dialect
+from sqlalchemy.engine.interfaces import Dialect
 
 from . import is_vertex_field_name
 
 
+# Describes the intent to join two tables using the specified columns.
+#
+# The resulting join expression could be something like:
+# JOIN origin_table.from_column = destination_table.to_column
+#
+# The type of join (inner vs left, etc.) is not specified.
+# The tables are not specified.
+DirectJoinDescriptor = namedtuple(
+    "DirectJoinDescriptor",
+    (
+        "from_column",  # The column in the source table we intend to join on.
+        "to_column",  # The column in the destination table we intend to join on.
+    ),
+)
+
+
+@dataclass
+class GenericSchemaInfo:
+    """Class for storing generic schema info required for querying."""
+
+    schema: GraphQLSchema
+
+    # Optional dict of GraphQL interface or type -> GraphQL union.
+    # Used as a workaround for GraphQL's lack of support for
+    # inheritance across "types" (i.e. non-interfaces), as well as a
+    # workaround for Gremlin's total lack of inheritance-awareness.
+    # The key-value pairs in the dict specify that the "key" type
+    # is equivalent to the "value" type, i.e. that the GraphQL type or
+    # interface in the key is the most-derived common supertype
+    # of every GraphQL type in the "value" GraphQL union.
+    # Recursive expansion of type equivalence hints is not performed,
+    # and only type-level correctness of this argument is enforced.
+    # See README.md for more details on everything this parameter does.
+    # *****
+    # Be very careful with this option, as bad input here will
+    # lead to incorrect output queries being generated.
+    # *****
+    # TODO: make sure we are treating empty dict the same as None
+    type_equivalence_hints: Optional[Dict[str, str]]  # Optional
+
+
+@dataclass
+class MatchSchemaInfo:
+    generic_schema_info: GenericSchemaInfo
+
+
+def create_match_schema_info(
```

Comment:
Discussed offline, but restating it here. 
I don't think that we want to expose any of these `create_*_schema_info` functions as API functions. It is too difficult to come up with the input. Instead we probably want to use `get_sqalchemy_schema_info` and `get_graphql_schema_from_orientdb_schema_data`, though those functions could probably use some refactoring themselves.

Reply:
Why not expose them? I agree that most users would just want to auto-infer their database's schema — that means we should expose the functions for that, and I don't think it means we shouldn't have these helper functions as well. I don't see any harm in exposing these, and I can think of use cases where we'd want them.

Reply:
I was hoping to expose those additional helper functions to help the user form the necessary input. They can then use those results with these helper functions.

Reply:
Easy interfaces are good for quick-start, modular interfaces are good for serious projects (immediately after quick-start). The only down side of exposing something is that we lock ourselves into keeping it stable, which should be the case for the schema info classes.

Reply:
I still don't think we should be exposing `create_mysql_schema_info`, `create_postgresql_schema_info`, etc. These functions take both `schema`and `vertex_name_to_table` as input. These two arguments both describe the schema of the underlying database and are therefore extremely correlated. Having two highly pieces of correlated data as arguments to a function is something that is prone to introduce bugs.

Reply:
`get_sqlalchemy_schema_info` does not have this issue since it generates the schema from `vertex_name_to_table` and `direct_edges`

Reply:
Side note: We now have `get_sqlalchemy_schema_info`, `make_sqlalchemy_schema_info` and `create_sqlalchemy_schema_info`.

Reply:
Agreed on the naming issues. As for whether these functions are exposed, I think they should be.

These functions are not intended as the "default" approach — the default is to create the schema info from an existing database. Calling these functions directly is effectively the power-user "I know what I'm doing" switch, and the documentation should indicate that misuse of these functions can produce a very broken compiler state.

We should do our best to validate that the inputs make sense (e.g. that vertex names in `vertex_name_to_table` match the schema types). But I don't think we should take away power from our power users just because that power could be misused to cause bugs.

Reply:
That's a good point - those names are pretty confusing.

Looking at it now, we shouldn't be exposing such similarly named functions (especially since they have very similar signatures). How about we prioritize the most obvious one to use? Would love to hear your thoughts about which one it is as I haven't tried using either (though I should!).